// @nrwl/workspace/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var typescript_1 = require("./src/utils/typescript");
exports.readTsConfig = typescript_1.readTsConfig;
var name_utils_1 = require("./src/utils/name-utils");
exports.toPropertyName = name_utils_1.toPropertyName;
exports.toClassName = name_utils_1.toClassName;
exports.toFileName = name_utils_1.toFileName;
exports.names = name_utils_1.names;
exports.findModuleParent = name_utils_1.findModuleParent;
var project_type_1 = require("./src/utils/project-type");
exports.ProjectType = project_type_1.ProjectType;
exports.projectRootDir = project_type_1.projectRootDir;
var fileutils_1 = require("./src/utils/fileutils");
exports.serializeJson = fileutils_1.serializeJson;
exports.renameSync = fileutils_1.renameSync;
exports.updateJsonFile = fileutils_1.updateJsonFile;
exports.readJsonFile = fileutils_1.readJsonFile;
exports.readWorkspaceConfigPath = fileutils_1.readWorkspaceConfigPath;
exports.copyFile = fileutils_1.copyFile;
exports.createDirectory = fileutils_1.createDirectory;
var common_1 = require("./src/utils/common");
exports.offsetFromRoot = common_1.offsetFromRoot;
exports.resolveUserExistingPrettierConfig = common_1.resolveUserExistingPrettierConfig;
var output_1 = require("./src/utils/output");
exports.output = output_1.output;
var nx_commands_1 = require("./src/command-line/nx-commands");
exports.commandsObject = nx_commands_1.commandsObject;
var supported_nx_commands_1 = require("./src/command-line/supported-nx-commands");
exports.supportedNxCommands = supported_nx_commands_1.supportedNxCommands;
var file_utils_1 = require("./src/core/file-utils");
exports.readWorkspaceJson = file_utils_1.readWorkspaceJson;
exports.readNxJson = file_utils_1.readNxJson;
var ast_utils_1 = require("./src/utils/ast-utils");
exports.readJsonInTree = ast_utils_1.readJsonInTree;
exports.updateJsonInTree = ast_utils_1.updateJsonInTree;
exports.updateWorkspaceInTree = ast_utils_1.updateWorkspaceInTree;
exports.insert = ast_utils_1.insert;
exports.replaceNodeValue = ast_utils_1.replaceNodeValue;
exports.addDepsToPackageJson = ast_utils_1.addDepsToPackageJson;
exports.addMethod = ast_utils_1.addMethod;
exports.addIncludeToTsConfig = ast_utils_1.addIncludeToTsConfig;
exports.addGlobal = ast_utils_1.addGlobal;
exports.getProjectConfig = ast_utils_1.getProjectConfig;
exports.addParameterToConstructor = ast_utils_1.addParameterToConstructor;
exports.createOrUpdate = ast_utils_1.createOrUpdate;
exports.findNodes = ast_utils_1.findNodes;
exports.updatePackageJsonDependencies = ast_utils_1.updatePackageJsonDependencies;
exports.getProjectGraphFromHost = ast_utils_1.getProjectGraphFromHost;
exports.readWorkspace = ast_utils_1.readWorkspace;
exports.renameSyncInTree = ast_utils_1.renameSyncInTree;
exports.renameDirSyncInTree = ast_utils_1.renameDirSyncInTree;
exports.updateNxJsonInTree = ast_utils_1.updateNxJsonInTree;
exports.addProjectToNxJsonInTree = ast_utils_1.addProjectToNxJsonInTree;
exports.readNxJsonInTree = ast_utils_1.readNxJsonInTree;
var cli_config_utils_1 = require("./src/utils/cli-config-utils");
exports.getNpmScope = cli_config_utils_1.getNpmScope;
exports.getWorkspacePath = cli_config_utils_1.getWorkspacePath;
exports.replaceAppNameWithPath = cli_config_utils_1.replaceAppNameWithPath;
exports.editTarget = cli_config_utils_1.editTarget;
exports.parseTarget = cli_config_utils_1.parseTarget;
exports.serializeTarget = cli_config_utils_1.serializeTarget;
var workspace_1 = require("./src/utils/workspace");
exports.getWorkspace = workspace_1.getWorkspace;
exports.updateWorkspace = workspace_1.updateWorkspace;
var update_task_1 = require("./src/utils/update-task");
exports.addUpdateTask = update_task_1.addUpdateTask;
var lint_1 = require("./src/utils/lint");
exports.addLintFiles = lint_1.addLintFiles;
exports.generateProjectLint = lint_1.generateProjectLint;
var add_install_task_1 = require("./src/utils/rules/add-install-task");
exports.addInstallTask = add_install_task_1.addInstallTask;
var format_files_1 = require("./src/utils/rules/format-files");
exports.formatFiles = format_files_1.formatFiles;
var deleteFile_1 = require("./src/utils/rules/deleteFile");
exports.deleteFile = deleteFile_1.deleteFile;
tslib_1.__exportStar(require("./src/utils/rules/ng-add"), exports);
var update_karma_conf_1 = require("./src/utils/rules/update-karma-conf");
exports.updateKarmaConf = update_karma_conf_1.updateKarmaConf;
var workspace_2 = require("./src/utils/rules/workspace");
exports.setDefaultCollection = workspace_2.setDefaultCollection;
const strings = require("./src/utils/strings");
var version_utils_1 = require("./src/utils/version-utils");
exports.checkAndCleanWithSemver = version_utils_1.checkAndCleanWithSemver;
var update_packages_in_package_json_1 = require("./src/utils/update-packages-in-package-json");
exports.updatePackagesInPackageJson = update_packages_in_package_json_1.updatePackagesInPackageJson;
exports.stringUtils = strings;
