// @nrwl/workspace/src/command-line/report.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular-devkit/core");
const fs_1 = require("fs");
const path = require("path");
const app_root_1 = require("../utils/app-root");
const output_1 = require("../utils/output");
exports.packagesWeCareAbout = [
    '@nrwl/angular',
    '@nrwl/cli',
    '@nrwl/cypress',
    '@nrwl/eslint-plugin-nx',
    '@nrwl/express',
    '@nrwl/jest',
    '@nrwl/linter',
    '@nrwl/nest',
    '@nrwl/next',
    '@nrwl/node',
    '@nrwl/react',
    '@nrwl/schematics',
    '@nrwl/tao',
    '@nrwl/web',
    '@nrwl/workspace',
    'typescript',
];
exports.report = {
    command: 'report',
    describe: 'Reports useful version numbers to copy into the Nx issue template',
    builder: (yargs) => yargs,
    handler: reportHandler,
};
/**
 * Reports relevant version numbers for adding to an Nx issue report
 *
 * @remarks
 *
 * Must be run within an Nx workspace
 *
 */
function reportHandler() {
    const nodeModulesDir = path.join(app_root_1.appRootPath, 'node_modules');
    const bodyLines = [];
    exports.packagesWeCareAbout.forEach((p) => {
        let status = 'Not Found';
        try {
            const packageJson = JSON.parse(fs_1.readFileSync(path.join(nodeModulesDir, p, 'package.json')).toString());
            status = packageJson.version;
        }
        catch (_a) { }
        bodyLines.push(`${core_1.terminal.green(p)} : ${core_1.terminal.bold(status)}`);
    });
    output_1.output.log({
        title: 'Report complete - copy this into the issue template',
        bodyLines,
    });
}
