// @nrwl/jest/src/builders/jest/jest.impl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const architect_1 = require("@angular-devkit/architect");
const jest_1 = require("jest");
const path = require("path");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
try {
    require('dotenv').config();
}
catch (e) { }
if (process.env.NODE_ENV == null || process.env.NODE_ENV == undefined) {
    process.env.NODE_ENV = 'test';
}
exports.default = architect_1.createBuilder(run);
function run(options, context) {
    options.jestConfig = path.resolve(context.workspaceRoot, options.jestConfig);
    const jestConfig = require(options.jestConfig);
    let transformers = Object.values(jestConfig.transform || {});
    if (transformers.includes('babel-jest') && transformers.includes('ts-jest')) {
        throw new Error('Using babel-jest and ts-jest together is not supported.\n' +
            'See ts-jest documentation for babel integration: https://kulshekhar.github.io/ts-jest/user/config/babelConfig');
    }
    // use ts-jest by default
    const globals = jestConfig.globals || {};
    if (!transformers.includes('babel-jest')) {
        const tsJestConfig = {
            tsConfig: path.resolve(context.workspaceRoot, options.tsConfig),
        };
        // TODO: This is hacky, We should probably just configure it in the user's workspace
        // If jest-preset-angular is installed, apply settings
        try {
            require.resolve('jest-preset-angular');
            Object.assign(tsJestConfig, {
                stringifyContentPathRegex: '\\.(html|svg)$',
                astTransformers: [
                    'jest-preset-angular/build/InlineFilesTransformer',
                    'jest-preset-angular/build/StripStylesTransformer',
                ],
            });
        }
        catch (e) { }
        // merge the jestConfig globals with our 'ts-jest' override
        Object.assign(globals, {
            'ts-jest': Object.assign(Object.assign({}, (globals['ts-jest'] || {})), tsJestConfig),
        });
    }
    const config = {
        _: [],
        config: options.config,
        coverage: options.codeCoverage,
        bail: options.bail,
        ci: options.ci,
        color: options.color,
        detectOpenHandles: options.detectOpenHandles,
        json: options.json,
        maxWorkers: options.maxWorkers,
        onlyChanged: options.onlyChanged,
        outputFile: options.outputFile,
        passWithNoTests: options.passWithNoTests,
        runInBand: options.runInBand,
        showConfig: options.showConfig,
        silent: options.silent,
        testLocationInResults: options.testLocationInResults,
        testNamePattern: options.testNamePattern,
        testPathPattern: options.testPathPattern,
        colors: options.colors,
        verbose: options.verbose,
        coverageReporters: options.coverageReporters,
        coverageDirectory: options.coverageDirectory,
        testResultsProcessor: options.testResultsProcessor,
        updateSnapshot: options.updateSnapshot,
        useStderr: options.useStderr,
        watch: options.watch,
        watchAll: options.watchAll,
        globals: JSON.stringify(globals),
    };
    if (options.setupFile) {
        config.setupFilesAfterEnv = [
            path.resolve(context.workspaceRoot, options.setupFile),
        ];
    }
    if (options.testFile) {
        config._.push(options.testFile);
    }
    if (options.findRelatedTests) {
        const parsedTests = options.findRelatedTests
            .split(',')
            .map((s) => s.trim());
        config._.push(...parsedTests);
        config.findRelatedTests = true;
    }
    if (options.clearCache) {
        config.clearCache = true;
    }
    if (options.reporters && options.reporters.length > 0) {
        config.reporters = options.reporters;
    }
    return rxjs_1.from(jest_1.runCLI(config, [options.jestConfig])).pipe(operators_1.map((results) => {
        return {
            success: results.results.success,
        };
    }));
}
