// @nrwl/angular/src/schematics/library/library.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const workspace_1 = require("@nrwl/workspace");
const init_1 = require("../init/init");
const add_module_1 = require("./lib/add-module");
const normalize_options_1 = require("./lib/normalize-options");
const update_lib_package_npm_scope_1 = require("./lib/update-lib-package-npm-scope");
const update_project_1 = require("./lib/update-project");
const update_tsconfig_1 = require("./lib/update-tsconfig");
function default_1(schema) {
    return (host) => {
        const options = normalize_options_1.normalizeOptions(host, schema);
        if (!options.routing && options.lazy) {
            throw new Error(`routing must be set`);
        }
        return schematics_1.chain([
            workspace_1.addLintFiles(options.projectRoot, "tslint" /* TsLint */, { onlyGlobal: true }),
            init_1.addUnitTestRunner(options),
            schematics_1.externalSchematic('@schematics/angular', 'library', {
                name: options.name,
                prefix: options.prefix,
                style: options.style,
                entryFile: 'index',
                skipPackageJson: !options.publishable,
                skipTsConfig: true,
            }),
            schematics_1.move(options.name, options.projectRoot),
            update_project_1.updateProject(options),
            update_tsconfig_1.updateTsConfig(options),
            options.unitTestRunner === 'jest'
                ? schematics_1.externalSchematic('@nrwl/jest', 'jest-project', {
                    project: options.name,
                    setupFile: 'angular',
                    supportTsx: false,
                    skipSerializers: false,
                })
                : schematics_1.noop(),
            options.unitTestRunner === 'karma'
                ? schematics_1.schematic('karma-project', {
                    project: options.name,
                })
                : schematics_1.noop(),
            options.publishable ? update_lib_package_npm_scope_1.updateLibPackageNpmScope(options) : schematics_1.noop(),
            add_module_1.addModule(options),
            workspace_1.formatFiles(options),
        ]);
    };
}
exports.default = default_1;
