// @nrwl/jest/src/schematics/jest-project/lib/update-workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular-devkit/core");
const workspace_1 = require("@nrwl/workspace");
function updateWorkspace(options) {
    return workspace_1.updateWorkspaceInTree((json) => {
        const projectConfig = json.projects[options.project];
        projectConfig.architect.test = {
            builder: '@nrwl/jest:jest',
            options: {
                jestConfig: core_1.join(core_1.normalize(projectConfig.root), 'jest.config.js'),
                tsConfig: core_1.join(core_1.normalize(projectConfig.root), 'tsconfig.spec.json'),
                passWithNoTests: true,
            },
        };
        if (options.setupFile !== 'none') {
            projectConfig.architect.test.options.setupFile = core_1.join(core_1.normalize(projectConfig.root), 'src/test-setup.ts');
        }
        if (projectConfig.architect.lint) {
            projectConfig.architect.lint.options.tsConfig = [
                ...projectConfig.architect.lint.options.tsConfig,
                core_1.join(core_1.normalize(projectConfig.root), 'tsconfig.spec.json'),
            ];
        }
        return json;
    });
}
exports.updateWorkspace = updateWorkspace;
