// storybook-addon-i18n/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var addons_1 = __importStar(require("@storybook/addons"));
var React = __importStar(require("react"));
var Provider_1 = __importDefault(require("./Provider"));
var shared_1 = require("./shared");
exports.withI18n = addons_1.makeDecorator({
    name: "withI18n",
    parameterName: "i18n",
    skipIfNoParametersOrOptions: true,
    allowDeprecatedUsage: true,
    wrapper: function (getStory, context, _a) {
        var parameters = _a.parameters;
        var channel = addons_1.default.getChannel();
        channel.emit(shared_1.ADDON_ID + "/register", parameters);
        return (React.createElement(Provider_1.default, __assign({ channel: channel }, parameters), getStory(context)));
    }
});
if (module && module.hot && module.hot.decline) {
    module.hot.decline();
}
